service/history/execution/mutable_state_util_test.go

```diff
@@ -316,3 +327,252 @@ func TestConvertWorkflowRequests(t *testing.T) {
 
 	assert.ElementsMatch(t, expectedOutputs, convertWorkflowRequests(inputs))
 }
+
+func TestCreatePersistenceMutableState(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	mockShardContext := shard.NewMockContext(ctrl)
+	logger := testlogger.New(t)
+	mockEventsCache := events.NewMockCache(ctrl)
+	mockDomainCache := cache.NewMockDomainCache(ctrl)
+	mockShardContext.EXPECT().GetClusterMetadata().Return(constants.TestClusterMetadata).Times(2)
+	mockShardContext.EXPECT().GetEventsCache().Return(mockEventsCache)
+	mockShardContext.EXPECT().GetConfig().Return(config.NewForTest())
+	mockShardContext.EXPECT().GetTimeSource().Return(clock.NewMockedTimeSource())
+	mockShardContext.EXPECT().GetMetricsClient().Return(metrics.NewClient(tally.NoopScope, metrics.History))
+	mockShardContext.EXPECT().GetDomainCache().Return(mockDomainCache)
+
+	builder := newMutableStateBuilder(mockShardContext, logger, constants.TestLocalDomainEntry)
+	builder.pendingActivityInfoIDs[0] = &persistence.ActivityInfo{}
+	builder.pendingTimerInfoIDs["some-key"] = &persistence.TimerInfo{}
+	builder.pendingSignalInfoIDs[0] = &persistence.SignalInfo{}
+	builder.pendingChildExecutionInfoIDs[0] = &persistence.ChildExecutionInfo{}
+	builder.bufferedEvents = []*types.HistoryEvent{{}}
+	builder.updateBufferedEvents = []*types.HistoryEvent{{}}
+	builder.versionHistories = &persistence.VersionHistories{CurrentVersionHistoryIndex: 0, Histories: []*persistence.VersionHistory{}}
+	builder.pendingRequestCancelInfoIDs[0] = &persistence.RequestCancelInfo{}
+	builder.decisionTaskManager.(*mutableStateDecisionTaskManagerImpl).HasInFlightDecision()
+	builder.executionInfo.DecisionStartedID = 1
+
+	mutableState := CreatePersistenceMutableState(builder)
```

Comment:
This is a mapper, taking a type (mutableStateBuilder) and creating a persistence Mutable State. Is it possible to see if there's the reverse somewhere? 

This kind of code is quite prone to not getting updated, I feel fairly strongly that we should round-trip test these mappings rather than just unit-testing them. Happy to sync to discuss what I mean. 

See `TestMutableStateBuilder_CopyToPersistence_roundtrip` for an example of a roundtrip test.

Reply:
the `msb.CopyToPersistance()` method I was testing looks weirdly similar, but I think they're different. Not super clear, but I think this one is a deep copy, so modifications to the copied persistence struct shouldn't affect the original maybe? not super sure how rigorous that is.

Reply:
Sure, I can add the roundtrip test on a follow up PR. I'm looking into the example.